src/theory/evaluator.cpp

```diff
@@ -290,10 +290,11 @@ EvalResult Evaluator::evalInternal(
           }
           ptrdiff_t pos = std::distance(args.begin(), it);
           currNodeVal = vals[pos];
+          needsReconstruct = false;
           // Don't need to rewrite since range of substitution should already
           // be normalized.
         }
-        else
+        if (needsReconstruct)
```

Comment:
Why is this necessary? In line 270 this `needsReconstruct` is set to true. If the if case (line 281) is executed it is now set to false. Therefore, the former else case is not executed. If the if case is not executed then the variable is still true. I might overlook a loop or something.